### DISCUSSION TITLE
Toolchain resolution with bzlmod, and toolchains registered in dependencies

### DISCUSSION THREAD

**Discussion by tpudlik** [bazelbuild/bazel#23075] [Q&A] [answered]
The [documentation](https://bazel.build/external/migration#register-toolchains) mentions that "toolchains and execution platforms registered by modules that are (transitive) dependencies of the root module" are considered during toolchain resolution, with lower precedence than toolchains registered in the root module itself. But,

1. What's the relative precedence of toolchains coming from two direct module dependencies?
2. Are toolchains from transitive dependencies lower precedence than toolchains from direct dependencies? (I believe https://github.com/bazelbuild/bazel/issues/20354#issuecomment-1831614509 proposed this, since it mentions precedence in BFS order, but was it implemented?)

@Wyverald

**Comment by Wyverald** [comment, marked as answer]
1. Among direct deps of the same module, earlier declarations win against later ones. That is, ordering of `bazel_dep` directives is significant in terms of toolchain resolution.

2. It's always BFS order

**Reply by tpudlik** [reply]
Ooh, I didn't realize the `bazel_dep` ordering mattered for anything, and happily sorted them alphabetically! Are there any circumstances other than toolchain resolution where their order matters?

**Reply by Wyverald** [reply]
Module extensions get a list of modules that use them (via `module_ctx.modules`). These are also sorted in BFS order (so root module first if it uses the extension, and then direct deps of root in source order, etc.)

I _think_ that's all.
